### DISCUSSION TITLE
BAPI_SALESORDER_CREATEFROMDAT

### DISCUSSION THREAD

**Discussion by ROLLS123** [dbosoft/YaNco#172] [Q&A] [answered]
![bapi call](https://user-images.githubusercontent.com/106260325/170853977-02e442a7-3db6-4f1f-a54d-9ee7df0305a3.png)
![BAPI CALL 2](https://user-images.githubusercontent.com/106260325/170853981-907731a0-c9a1-4eed-a5f2-69b97a5ed726.png)
this function create document in SAP can you look my codes what missing because not creating a document in SAP with no error..thank you

**Comment by fw2568** [comment]
Hi,

please note that we generally do not provide support for business use here. You can request commercial support on [https://support.dbosoft.eu](https://support.dbosoft.eu/hc/en-us/articles/360021454080-YaNco-Support-Plans) However, I will try to help you a little. 

If I understand you correctly, you are looking for a way to read the result of the BAPI call. You can do this if you use method CallFunction instead of CallFunctionOneWay. The one-way function is used for fire and forget calls without processing any output. If you use CallFunction you can process the response of the BAPI (table "return"). However, to get to the reasons why something is not working, in most cases it is better to set a remote breakpoint in the backend and observe how the data is being processed. 

I also noticed your calls to .Append(structure). If this is not some extension you created please not that it is not necessary to Append the structure to the table. The mapping function is called for each entry in the enumerable you pass to SetTable and each structure row is created automatically. So you only have to specify how to map from your internal data to the BAPI fields. 
Date fields are also handled automatically (DateTime or DateTimeOffset), so you don't have to parse them to SAP internal format. 

Hope that helps,

Best Regards,
Frank

**Reply by ROLLS123** [reply]
Hi fw2568 ,

Thank you for your tips and response . by the way I'm looking for SAP Doc Transfer or Create Sales order transaction using YANCO i just wonder to how to use commit function or BAPI_TRANSACTION_COMMIT inside callfunction below is my sample code.
![BAPI CALL 3](https://user-images.githubusercontent.com/106260325/171124749-4258d06f-5f13-4ae2-ac25-f56b9a8df618.png)
![BAPI CALL 4](https://user-images.githubusercontent.com/106260325/171124769-4cbdfa8c-cc67-41fc-8008-375a35a6c71d.png)

here is blog link  about create-sales-order in SAP.  https://blogs.sap.com/2016/01/21/create-sales-order-using-bapi-with-variant-configuration/
https://answers.sap.com/questions/3204971/bapisalesordercreatefromdat2-sample.html
Regards,
Rolls

**Comment by fw2568** [comment]
You always commit a session, so you have to use `context.CommitAndWait()`

**Reply by ROLLS123** [reply]
i putted commit after the call of function same issue no created data in SAP may i know where the right place to call commit and how to insert handle return in commit  in this code
below the result of commit.
![BAPI CALL 5](https://user-images.githubusercontent.com/106260325/171338844-88e16655-7e11-40e4-bd70-8d968dd709eb.png)

Thanks and Regards,
Rolls

**Reply by ROLLS123** [reply]
i also tried this but same result i hope you can help me for this f
![BAPI CALL 6](https://user-images.githubusercontent.com/106260325/171435768-17297027-a318-4ffa-aae0-8da1a55cfc0b.png)
unction.

**Reply by fw2568** [reply]
My answer about the commit was a bit incomplete. You have to commit the context, but there is also a helper to chain it with the CallFunction method. 
This is one of the fundamental concepts of functional programming - the result of each call is a Either<L,R> value and normally methods are only called while the Either is in Right state.
So If you call 
```
CallFunction(... )
.CommitWorkAndWait(context)
```

the commit logic will only be called when the Either is in right state. 
However to do this, you have to handle the result of the return table between the calls. I have updated examples and currently preparing a PR that shows this scenario.

If you don't map the result of the RETURN, you can do it like here:

```
var result = await CallFunction(...)
                   .MatchAsync( ... );

//check if result is ok or do some other work
...
// finally commit
await context.CommitAndWaitAsync();

```

Please also don't use Rollback for a failed call - this not required for BAPI calls, only in some special [cases](https://help.sap.com/doc/abapdocu_751_index_htm/7.51/en-us/abaprollback.htm).

BTW: please try to copy code instead of screenshots - screenshots are hard to read, to comment and not searchable.

**Comment by fw2568** [comment, marked as answer]
Please see the CreateSalesOrder sample in https://github.com/dbosoft/YaNco/tree/main/samples/netcore3.1/CreateSalesOrder for a example how to use YaNco to create a sales document. 

https://github.com/dbosoft/YaNco/blob/0bf4705b4992d67ac19bf54aa9058f6664911e39/samples/netcore3.1/CreateSalesOrder/CreateSimpleSalesDocument.cs#L43-L98

**Comment by ROLLS123** [comment]
Hi fw2568,

its work for my side perfect thank you for your help.
